doc/y_cable/design_doc.md

```diff
@@ -0,0 +1,226 @@
+
+This document summarizes the approach taken to accommodate multiple
+Y cable vendors for supporting dual-ToR configurations in SONiC.
+All the design choices are presented
+such that vendors can implement the Y cable interface with ease and effeciency
+and all the feature requirements are met.
+
+
+Challenge: to provide an interface for Y cable to interact with PMON docker and HOST
+           which can be uniform across all vendors.
+
+### Directory Structure
+
+#### Background
+
+- We need to have an appropriate place where vendors can implement their modules/packages and provide Sonic with an abstraction layer where pmon and host can access the Y cable packages of all the vendors with ease
+   
+  - Currently the assumption is Y cable package implementation would be in python itself
+  - If a vendor provides shared library, that scenario would be considered accordingly
+  - Since sonic_y_cable (sonic-platform-common/sonic_y_cable) is already built as a package today (only for a single Y cable vendor with sonic-buildimage) vendors can also build packages along similar lines.
+
+#### Proposed Solution
+
+- Vendors can place their implementations in this format
+
+```
+	Sonic_platform_common/sonic_y_cable/<vendor>/y_cable.py
+```
+- For sxample
+
+```
+	Sonic_platform_common/sonic_y_cable/credo/y_cable.py
+```
+#### Rationale
+- The requirement here would be that the vendors must create their packages such that It can easily be accessed/imported and the packages also adhere to a uniform convention
+
+  - Vendor name should be derivable from the vendor name field inside the register spec of transceiver type.
+  - This is essential so that host/pmon can determine the vendor name from eeprom and import the module of the transceiver appropriately
+
+### Vendor to Module mapping
+
+#### Background
+
+- We need to have a vendor to Y cable module mappings defined and accessible somewhere for pmon/cli to import
+   
+  - Before calling any Y cable API, how do we know which type/vendor does the Y cable belong to ? 
+  - Also lets suppose once the type/vendor of the cable is known from where does pmon or cli import the module ?
+  - Since there can be many specs which vendors follow (8636, 8624 etc.) for transceivers it is important to have a solution which meets all requirements so that the API access is available to whoever wants to invoke it.
+  - another issue is how do we support the multiple vendors modules/packages which could be used by both sonic pmon docker as well as the sonic cli (sonic-utilities). Basically the package should be available both in host and container.
+#### Proposed Solution
+
+- We define a file which would contain vendor to module mapping and put it in a place which is accessible to both pmon and host
+- The vendor to module mapping can be present in a file which could be present in
+
+```
+	/usr/share/sonic/y_cable/y_cable_vendor_mapping.csv
+```
+or 
+
+```
+	/usr/share/sonic/y_cable/y_cable_vendor_mapping.yaml
```

Comment:
We need to decide upon a data format here, rather than leaving it open.

I think it makes the most sense to keep the mapping file in the sonic_y_cable package so that it can be updated in the same pull request when a vendor adds a new cable implementation. However, we cannot install data files outside of the Python directory via a wheel. So maybe we need to make this a Python file which simply contains a 2D dictionary which can be used to look up the module path? If the file is part of the sonic_y_cable package, it will be installed in both the PMon container and the host.

Reply:
fixed